### DISCUSSION TITLE
"ignoring illegals" and "no raid" syntax questions

### DISCUSSION THREAD

**Discussion by esksp** [endless-sky/endless-sky#10662] [Q&A]
I've made a couple new governments and systems, and noticed they throw a number of **errors in the error log**.
Clearly I've missed something in using those attributes, but then again there is no example in the existing game files I could use to do my usual "monkey see, monkey do".

---

What is the exact syntax for **governments ignoring illegals**?
I can't find any "illegals [ignore]" attribute anywhere in the game files, yet I don't think pirates fine you for carrying nerve gas, do they? How is this currently handled?
This is one of those cases where the Wiki is way too vague. All it says is it should look like:

	illegals
		[ignore]  <outfit> <fine>

Why on earth does ignoring something require also declaring a fine? What should I put there?
In the meantime I tried to ignore the fine, and tried this entry ("Spice" is an illegal outfit (drug)):

	illegals
		[ignore] Spice
		[ignore] "Nerve Gas"

I wasn't yet able to check if this actually works in game, but I already know the error log doesn't like it. It says: "Cannot convert value "Spice" to a number", "Cannot convert value "Nerve Gas" to a number"! What am I to make of _that_? All I can say is "you're not supposed to"...
I would suspect it expects quotes around an outfit name, but then it doesn't like "Nerve Gas" either. Apparently it _so much_ wants a fine value, that if there is only a single entry it thinks it _must_ be the fine value, even if it's text.

So, what's the right syntax for saying a given government doesn't care about you carrying specific illegals?

---

I've got a similar problem with the **"no raid" attribute**, supposed to prevent raids: According to the error log the game expects a value. A value of what exactly? For me "no" means "no"...
The Wiki doesn't mention any value, and I don't see what value it might expect anyway.

---

And while we're at it, **the error log complains about a lot of stuff from the vanilla "map_systems.txt"**, with a slew of "**unrecognized attribute**" entries for things like "minables", "belt", "arrival", and several others.
Is the error parser up to date?

**Comment by TheGiraffe3** [comment]
For illegals ignore, are you using `[]` around `ignore`? You technically aren't supposed to do that.

---

Theoretically, you just don't define a raid, and you won't get one?

---
What version do you have?

**Reply by esksp** [reply]
> For illegals ignore, are you using `[]` around `ignore`? You technically aren't supposed to do that.

Really? I wrote it the only way I know, the way the Wiki says (see my examples above)...
Like I said, the Wiki is too succinct... :-(

So, what is the _real_ syntax for telling "this government doesn't mind Spice and Nerve gas"?

 
> Theoretically, you just don't define a raid, and you won't get one?

Really? Good, that's definitely easier!...


> What version do you have?

v0.10.8 (from GOG, on Linux).

**Reply by TheGiraffe3** [reply]
`[ignore] <outfit>`: if you remove the `[]` then that outfit is ignored. Otherwise you remove the whole word (to get `<outfit>`.

**Comment by tibetiroka** [comment]
> I've got a similar problem with the "no raid" attribute

Ah, the attribute is `"no raids"` in the code. I'll update the wiki.

**Comment by tibetiroka** [comment]
> And while we're at it, the error log complains about a lot of stuff from the vanilla "map_systems.txt", with a slew of "unrecognized attribute" entries for things like "minables", "belt", "arrival", and several others.
Is the error parser up to date?

Weird. Arrival distances were added in 0.9.17, minables have been there since forever. Can you post your errors.txt file?

Btw, there is no dedicated error parser for these. It just gives you an error when it doesn't recognize the meaning; so if you get an error about vanilla content, it's usually a sing of the data being newer than the game version.

**Reply by esksp** [reply]
> Weird. Arrival distances were added in 0.9.17, minables have been there since forever. Can you post your errors.txt file?

Sure. Since there were some entries from previous iterations of my plugin in there, I deleted it about an hour ago to clean it up.
But I just started ES, and of course it created a new one - With the same old/orphan entries!

And an UFO: Somewhere near the end the error log states:
```
(Spicehead Raven):
outfit space: -1
has outfits:
	1 Scram Drive
	1 Liquid Nitrogen Cooler
	1 Outfit Scanner
	1 Cargo Scanner
	4 Particle Cannon
	1 Large Radar Jammer
	1 Water Coolant System
	1 A250 Atomic Thruster
	1 A255 Atomic Steering
	1 Fission Reactor
	1 LP576a Battery Pack
```

And yet the actual "Spicehead Raven" ship has an outfit space of 400.

Here is its actual definition (it's a copy-paste variant of the standard Raven with no shields but better engines and equipment, which is why it is "uncapturable"):
```
ship "Spicehead Raven"
	sprite "ship/raven"
	thumbnail "thumbnail/raven"
	attributes
		category "Light Warship"
		"cost" 2200000
		"uncapturable"
		"shields" 2400
		"hull" 1400
		"required crew" 60
		"bunks" 80
		"mass" 100
		"drag" 3.7
		"heat dissipation" .9
		"fuel capacity" 800
		"cargo space" 30
		"outfit space" 400
		"weapon capacity" 160
		"engine capacity" 310
		weapon
			"blast radius" 60
			"shield damage" 600
			"hull damage" 300
			"hit force" 900
	outfits
		"Particle Cannon" 4
		"Large Radar Jammer"
		"Cargo Scanner"
		"Outfit Scanner"
		"Fission Reactor"
		"Water Coolant System"
		"Liquid Nitrogen Cooler"
		"LP576a Battery Pack"
		"A250 Atomic Thruster"
		"A255 Atomic Steering"
		"Scram Drive"
	engine -9.5 63
	engine 9.5 63
	gun -13.5 -35.5 "Particle Cannon"
	gun 13.5 -35.5 "Particle Cannon"
	gun -21 -28 "Particle Cannon"
	gun 21 -28 "Particle Cannon"
	leak "leak" 60 50
	leak "flame" 50 80
	explode "medium explosion" 24
	explode "large explosion" 16
	explode "tiny explosion" 28
	explode "small explosion" 40
	"final explode" "final explosion small"
	description "This is a customized Raven. You won't survive long on it since most safety and life support systems have been stripped, but it is fast, agile and has serious fire power."

````

What does that "outfit space: -1" mean? And why don't the other ships (there are a couple similar variants) don't throw a similar error?
Also, the "uncapturable" attribute of some of those variants seems to throw an error. And yet, it's a copy/paste from the Wiki...

Anyway, here is the error log (zipped because it's long! 1531lines, 4.9 MB!)

[error.txt.zip](https://github.com/user-attachments/files/17451295/error.txt.zip)

In the end I added some comments so you know which entries are due to my plugin and which are due to the default data.

> it's usually a sing of the data being newer than the game version.

How can the data be newer than the game version it came with? It's a fresh installation from two weeks ago. Only modification is the plugin I'm working on, but that stays inside /plugins, I don't touch the original files.

**Reply by danaris** [reply]
> How can the data be newer than the game version it came with?

The most common way I run into it is when I have multiple versions of the game on my computer, and I accidentally start up an older one without realizing it (and it's accessing data from/for a newer one, usually in the plugin directory).

**Reply by esksp** [reply]
Yeah but no, I only have this one...
(Well, I also have a version installed on a Window box to get to the data, since my Linux version is an Appimage and I can't access the actual files. But this is irrelevant, different box, different OS, and besides the Windows copy is the same version (0.10.8).)

No, my game is pretty much whatever I downloaded from GOG two weeks ago. I only discovered ES two weeks ago...

**Reply by TheGiraffe3** [reply]
> error.txt.zip

It looks like you're redefining vanilla systems, and have got the indentation wrong. Most of your errors file is that. Do you need to redefine them all? Would probably be better not to.
The `uncapturable` attribute should be right under `thumbnail`, instead of in the `attributes` node.
You should remove `[]` from `[ignore]`, so that it reads `ignore Spice`.
Change `no raid` to `no raids`.

That should fix almost everything.
Then try starting a new save for testing, and most other errors should be gone. Could you upload your errors file after that (if there's anything left besides `Warning: sound "" is referred to, but not fully defined)?

**Reply by esksp** [reply]
> It looks like you're redefining vanilla systems, and have got the indentation wrong.

This was first created by "**endless-sky-editor-win64-0.7.10-2**", which I used to add new systems.
It loads the whole "systems" file, and then you save it as a new file.
I then manually edited that new file to add the system specifics (the editor only does place stars and planets). I intend to eventually erase from that file all the systems which were not actually modified (a couple were modified, to add links to the new ones), but for the time being I kept it as is, a full, modified copy of the original "map systems.txt".

> The `uncapturable` attribute should be right under `thumbnail`, instead of in the `attributes` node.
> You should remove `[]` from `[ignore]`, so that it reads `ignore Spice`. Change `no raid` to `no raids`.

Done, done, done. Thanks.

> That should fix almost everything. Then try starting a new save for testing, and most other errors should be gone.

```
Error: Expected key to have a value:
file /home/esksp/.local/share/endless-sky/plugins/StrikeForces/data/government.txt
L1:   government "Strike Forces"
L41:     "no raids"

```
I can't get it right, do I... :-D
Oh well, I'll remove that attribute entirely, but apparently "no raids" isn't recognized either.

Here is the new error log (fresh, I deleted the old one once again). It's still 4.9 MB, _some_ entries did disappear, but not many... I still get all those "minables", "belt", "arrival" and other errors.
[error.txt2.zip](https://github.com/user-attachments/files/17451468/error.txt2.zip)

**Reply by tibetiroka** [reply]
`"no raids"` should go in the system, not the government, I think.

**Reply by esksp** [reply]
I made the final test, I removed my plugin, started ES, and lo and behold, the error log was (almost) empty.

So it's indeed my "map systems.txt" modified by "endless-sky-editor-win64-0.7.10-2" and then by me which causes the problems.
Go figure why.
And how I can fix it (it's many days of work, including new 3D models and stuff. I'd hate to throw it away).  :-(

**Reply by warp-core** [reply]
Can you uplaod your modified map file? It looks not good from the errors file.
In general, it's best not to try to modify the original game files for creating a plugin. A plugin can update existing items (such as planets and systems) without copying the original definition.

**Reply by TheGiraffe3** [reply]
> I made the final test, I removed my plugin, started ES, and lo and behold, the error log was (almost) empty.
> 
> So it's indeed my "map systems.txt" modified by "endless-sky-editor-win64-0.7.10-2" and then by me which causes the problems. Go figure why. And how I can fix it (it's many days of work, including new 3D models and stuff. I'd hate to throw it away). :-(

I would recommend using quyykk's editor instead of the old one. It does not save the whole file, and has many more features, including the ability to generate system insides.
https://github.com/quyykk/plugin-editor

**Reply by esksp** [reply]
> Can you uplaod your modified map file?

Sure, here you go (systems and planets). It's my working files:
[map_data.zip](https://github.com/user-attachments/files/17459230/map_data.zip)


>In general, it's best not to try to modify the original game files for creating a plugin. A plugin can update existing items (such as planets and systems) without copying the original definition.

I understand, but that was the editor's choice, take an exiting file and add modifications, so I went with it.
I kept the idea of a single, complete systems file, because it's easier to have everything in one place while working (and because said editor would only open one file, and I needed to know what else is there).

That been said, I _did_ suspect the editor might damage the original systems' information, since it seemed a little outdated/simplified. So, once it gave me a modified systems file, **I just copy-pasted the modified/new systems into a copy of the original "map systems.txt" file**, to limit damages. It's the one I gave you.
Meaning, the only _original_ systems which _could_ have been damaged by me, are the 4-5 to which I added links (although I just copied the "link" lines into them, so it's rather unlikely. I never touched "minables" or such).
My point is, **it is strange I get errors for things I normally didn't touch**. Go figure. Unfortunately I'm busy today, I probably won't have the time to investigate this further before tomorrow.

**Reply by esksp** [reply]
> I would recommend using quyykk's editor instead of the old one.

Yes, thanks, but I won't start over. But I will definitely use it the next time...  :-)

Besides, as I wrote above (below, sideways, I haven't got the hang of GitHub yet):
I _did_ suspect the editor might damage the original systems' information, since it seemed a little outdated/simplified. So, once it gave me a modified systems file, **I just copy-pasted the modified/new systems into a copy of the original "map systems.txt" file**, to limit damages. It's the one I posted above.
Meaning, the only _original_ systems which _could_ have been damaged by me, are the 4-5 to which I added links (although I just copied the "link" lines into them, so it's rather unlikely. I never touched "minables" or such).
My point is, **it is strange I get errors for things I normally didn't touch**. Go figure. Unfortunately I'm busy today, I probably won't have the time to investigate this further before tomorrow.

**Reply by TheGiraffe3** [reply]
If that's the only change you're trying to make, it's easy to fix.
All you do is:
`system <name of the system>`
<tab>`add link <name of your system>`.
Take this line for the Jibauni as an example: https://github.com/TheGiraffe3/Jibauni/blob/05ac971f63183e4f53955a4245247ff89e0e2ad4/data/map%20systems.txt#L1069

**Reply by TheGiraffe3** [reply]
> Yes, thanks, but I won't start over.

You don't have to. Both of them read plugin data the same way, so there won't actually be any compatibility issues.
The only problem you may run into is that the Ship Editor crashes quyykk's editor, but you don't have to use that element.

**Reply by esksp** [reply]
> If that's the only change you're trying to make, it's easy to fix. All you do is: `system <name of the system>` `add link <name of your system>`.

So easy?
Well, if only I had known it...

Risking to sound like Cato the Elder, the Wiki is too vague and lacks information...
That detail should be front and center on the  "Creating Systems and Planets" page. Most plugin makers just want to _add_ stuff, not start from scratch, so the commands to add/modify stuff are the most important. Yet there is no mention of them. :-(

**Reply by TheGiraffe3** [reply]
It mentions being able to do that here: https://github.com/endless-sky/endless-sky/wiki/CreatingPlugins#creating-or-overriding-data-elements
But yes, it is somewhat vague.

**Reply by esksp** [reply]
> You don't have to. Both of them read plugin data the same way, so there won't actually be any compatibility issues.

The thing I'm afraid of is that the editor might not recognize some (newer) attributes, and just silently remove/damage them without me noticing. I don't know enough to spot such things, so I try to play it safe.
For this reason I only used the (other) editor to do what it does best (and I can't do manually): Place systems/planets in a logical way at logical places. And then I added the rest manually in a text editor.

> but you don't have to use that element.

No, indeed, ships are pretty straightforward. The only thing not doable by hand is placing the guns/engines, but there is the online editor for that.
As for calculating outfits, I just make the ships temporarily buyable and use the Outfitter in-game... (Most of my ships are not buyable/capturable, their only point is to keep the late-game player humble... -evil laughter-)

**Reply by esksp** [reply]
> It mentions being able to do that here: https://github.com/endless-sky/endless-sky/wiki/CreatingPlugins#creating-or-overriding-data-elements But yes, it is somewhat vague.

And hard to find! 
If I want to add some systems, I would tend to look in the "How to add systems" page, isn't it...
The "Creating plugins" page would be to explain the general plugin structure, and way plugins work (which information supersedes which, and stuff like that).
Illogical.
/Spock

**Reply by warp-core** [reply]
It looks like a lot (every?) system "object" has had the "attributes", "belt", "arrival", and "minable" nodes from its system copied into it.
These nodes should not be children of "object".
I can remove those erroneous lines for you shortly.

**Reply by esksp** [reply]
> It looks like a lot (every?) system "object" has had the "attributes", "belt", "arrival", and "minable" nodes from its system copied into it.
> These nodes should not be children of "object".

Sorry, but those are, just like that, in the default "systems" files which came with my game download.
I don't change things which should work... :-D
I guess it's a bug of version 0.10.8 (Windows). (To clarify, I had to download and unzip a Windows version to get to those files: The Linux download (GOG) is an Appimage file and I can't easily access the files.)

**Reply by TheGiraffe3** [reply]
Are they? Or are they the systems the editor saves?
Seeing as how the official editor is about 6 years out of date, I would assume it's the editor.

**Reply by warp-core** [reply]
That's definitely not what the vanilla files look like.
Here's the first system definition from 0.10.8 (with the list of objects truncated because we dont' need them all):
https://github.com/endless-sky/endless-sky/blob/d5dcdd6269d02dd6d0ef7fae0f4fe994936cfa6e/data/map%20systems.txt#L81-L130
Here's what I get for the same system from the file you uploaded:
```
system "1 Axis"
	pos -1274.63 267.214
	government Coalition
	habitable 745
	link "3 Axis"
	link "4 Spring Rising"
	link "4 Winter Rising"
	link "Sol Kimek"
	asteroids "small rock" 8 3.3166
	asteroids "medium rock" 11 3.4454
	asteroids "large rock" 1 3.1556
	asteroids "small metal" 1 3.381
	asteroids "medium metal" 149 4.347
	asteroids "large metal" 11 3.9606
	trade Clothing 251
	trade Electronics 831
	trade Equipment 469
	trade Food 231
	trade "Heavy Metals" 1241
	trade Industrial 767
	trade "Luxury Goods" 1135
	trade Medical 777
	trade Metal 520
	trade Plastic 384
	fleet "Small Arach" 700
	fleet "Large Arach" 1500
	fleet "Small Kimek" 300
	fleet "Large Kimek" 400
	fleet "Small Saryd" 900
	fleet "Large Saryd" 1800
	fleet "Arach Miners" 2500
	fleet "Kimek Miners" 2000
	fleet "Saryd Miners" 3000
	fleet Heliarch 450
	attributes kimek
	arrival 745
	belt 1112
	minables lead 11 3.87213
	minables silicon 35 3.68659
	minables tungsten 2 4.45307
	object
		sprite star/k5
		distance 45.3586
		period 15.903
		offset 180
		attributes kimek
		arrival 745
		belt 1112
		minables lead 11 3.87213
		minables silicon 35 3.68659
		minables tungsten 2 4.45307
	object
		sprite star/m0
		distance 60.2419
		period 15.903
		attributes kimek
		arrival 745
		belt 1112
		minables lead 11 3.87213
		minables silicon 35 3.68659
		minables tungsten 2 4.45307
	object
```

**Reply by esksp** [reply]
> Are they? Or are they the systems the editor saves?

No, it's what I see in **/data** after unzipping the **Windows ES version 0.10.8 (no letter) from GOG**. As downloaded. 
I never overwrite original files... That would be reckless, to stay polite.

> That's definitely not what the vanilla files look like.

I believe you!
It seems there is something fishy in the Windows GOG version 0.10.8 of the game.
Anyway, where can I get a copy of the correct /data (see #10666) to replace them?

**Reply by esksp** [reply]
**Downloaded a clean copy of 0.10.8, and indeed, a lot of files have different sizes**, not only systems and planets (also dialogues, governments, and I didn't even check the subfolders), even the main executable has a different size: **There is (was) a problem with the GOG Windows version**.

**Reply by esksp** [reply]
> **There is (was) a problem with the GOG Windows version**.

And I know what it is: There is a changelog in there, and it states **version 0.10.6**!...
Somebody mixed up their files when packaging... :-D

In the meantime GOG has an updated version for download, so I guess they have noticed.

**Reply by esksp** [reply]
> I would recommend using quyykk's editor instead of the old one. It does not save the whole file, and has many more features, including the ability to generate system insides. https://github.com/quyykk/plugin-editor

I've tried it out, but it displays a huge lot of new systems and pretends to be for 0.10._15_!
Yikes, that's 7 points more than I need...  :-(

Is there a way to make it use the 0.10._8_ files instead of the "continuous build" ones?

**Reply by danaris** [reply]
Well, 0.10.15 doesn't exist yet, and the "continuous build" release of the plugin editor that's available for download from GitHub was last updated in March—so it should be between 0.10.6 and 0.10.7.

Assuming there aren't disparities between what's actually in the data files and what that version can parse, you should be able to have the editor base its editing on whatever version you want simply by replacing the `data` folder it has with a copy of the one from the version of ES you're playing.

**Reply by esksp** [reply]
> Well, 0.10.15 doesn't exist yet, and the "continuous build" release of the plugin editor that's available for download from GitHub was last updated in March—so it should be between 0.10.6 and 0.10.7.

I don't know, I'm just stating what it says in the "About" menu...


> by replacing the `data` folder it has with a copy of the one from the version of ES you're playing.

True, but the problem is that all Linux versions (ES and this editor) are _Appimages_, i.e.  huge self-contained single files, on which you can't access the individual parts... The /data folder is hidden somewhere in this 243 MB file, and only accessible in RAM when you run it... 

I guess I'll have to find me a Windows computer to install it to, since the Linux version is pretty much unusable. :-(

**Reply by TheGiraffe3** [reply]
Not sure why it would be 0.10.15... is that the editor's version?
I know the version on my Mac isn't that.

Can you right-click on the app and do something like "Show Package Contents"?

**Reply by danaris** [reply]
> I don't know, I'm just stating what it says in the "About" menu...

You sure you're not misreading it saying it's for "v0._9_.15"? ^_^;;

There do appear to be ways to extract and re-pack an AppImage, but I'm not sure whether it would be easier to just get the source and rebuild it yourself...

**Reply by esksp** [reply]
> Not sure why it would be 0.10.15... is that the editor's version?

Oops, my bad! I misread, it actually says "for Endless Sky 0.**9**.15"... ("Built on 16 Mar 2024")
Still...

> Can you right-click on the app and do something like "Show Package Contents"?

No, Appimages are self-contained mini-VM systems which expand to a full container with its own filesystem when run.

Note I found an "**open game data (experimental)**" menu entry, so I'll experiment with that. If it doesn't blow my computer up...  :-D

**Reply by esksp** [reply]
> Note I found an "**open game data (experimental)**" menu entry

Nope. It did indeed allow me to open the right data, but hard froze when I tried to also load my (quite simple) plugin.

I also downloaded and tried the **Windows version, under Wine**: It seems to work, allowed me to load my plugin and add a very simple test system, but when I tried to save the result ("Save As"), the editor crashed (and didn't save anything)...
I'm afraid https://github.com/quyykk/plugin-editor isn't yet usable for Linux users. :-(

**Reply by TheGiraffe3** [reply]
That's sad.
Good luck with the old one!

**Comment by warp-core** [comment]
> What is the exact syntax for **governments ignoring illegals**? I can't find any "illegals [ignore]" attribute anywhere in the game files, yet I don't think pirates fine you for carrying nerve gas, do they? How is this currently handled? This is one of those cases where the Wiki is way too vague. All it says is it should look like:
> 
> ```
> illegals
> 	[ignore]  <outfit> <fine>
> ```
> 
> Why on earth does ignoring something require also declaring a fine? What should I put there? In the meantime I tried to ignore the fine, and tried this entry ("Spice" is an illegal outfit (drug)):

You're right, this ought to be two separate lines:
```
illegals
    <outfit> <fine>
    ignore <outfit>
```
